src/name.rs

```diff
@@ -0,0 +1,28 @@
+//! Support for associating type name information with a [`Message`].
+
+use crate::Message;
+use alloc::{format, string::String};
+
+/// Associate a type name with a [`Message`] type.
+pub trait Name: Message {
+    /// Type name for this [`Message`]. This is the camel case name,
+    /// e.g. `TypeName`.
+    const NAME: &'static str;
+
+    /// Package name this message type is contained in. They are domain-like
+    /// and delimited by `.`, e.g. `google.protobuf`.
+    const PACKAGE: &'static str;
+
+    /// Full name of this message type containing both the package name and
+    /// type name, e.g. `google.protobuf.TypeName`.
+    fn full_name() -> String {
```

Comment:
Should these be `&'static str`, is there a case where these are not statically defined with the generated proto?

Reply:
If `prost-build` filled them in they potentially could be all computed at build-time.

Otherwise it's not really possible to construct a `&'static str` as a concatenation of `Self::PACKAGE` + `'.'` + `Self::NAME` AFAIK.